### ISSUE TITLE
Code generated by Kotlin compiler plugins is not available when detekt compiles code for analysis

### ISSUE THREAD

**Issue by mgroth0** [detekt/detekt#7531]
## Expected Behavior

If a kotlin module compiles without error or warning (`-Werror` enabled), then the `Detekt` gradle task shouldn't have compiler errors.

## Observed Behavior

I get compiler errors when running the `Detekt` gradle task. I get a few different types of of errors, but the type I will focus on in this issue is `unresolved reference`.

## Steps to Reproduce

1. Create a kotlin module that uses the kotlinx.serialization plugin.
2. Write some code that uses the plugin-generated `serializer` function. This function is generated onto the companion object. For example, `MyClass.serializer()`. 
    - Note: in my case, I reproduce this in the `commonMain` source set in a multiplatform module.
4. Run the `Detekt` gradle task with debug mode enabled
5. Observe output resembling the following:

```log
/my/project/src/commonMain/kotlin/my/package.kt:56:55: error: unresolved reference: serializer
       MyClass.serializer()
                      ^^^^^^^^^^
There were 6 compiler errors found during analysis. This affects accuracy of reporting.
Run detekt CLI with --debug or set `detekt { debug = true }` in Gradle to see the error messages.
```

## Context

Previously, I was using an older version of Detekt. It was my own git branch, but contained commits up to about December 2023, including from 3flex's K2 branch.

Just recently, I upgraded. I discarded my custom commits and updated all the way up to the most recent commit of the `main` branch (https://github.com/detekt/detekt/commit/7675ae87250594f16184069f7ac2c300d86ed070). My upgrade included many major changes that were added in the last half year. Since the main branch includes kotlin 2 support and fixed the issue that I previously made a custom fix for, I no longer need my custom branch.

Generally, all works well after my upgrade. I can run the `Detekt` gradle task and am seeing various good improvements and fixes. However, I am getting these `compiler errors` everywhere. **These did not happen before (or at least were not reported before).**

In this issue I am focusing on just one type of compiler error, in hopes that the solution to it will lead to the solutions for others as well.

While I don't have a complete understanding about what is happening, it would seem that perhaps Detekt is using an embedded kotlin compiler which might be configured in a different way that is incompatible with my code (for example, maybe the embedded compiler somehow isn't aware of the serialization compiler plugin?). I've seen a few commits that seemed related to this. Perhaps this is an actual bug, or perhaps I am not configuring the `Detekt` task correctly with the latest updates.

## Your Environment

* Version of detekt used: Main branch, commit 7675ae87250594f16184069f7ac2c300d86ed070
* Version of Gradle used (if applicable): 8.8
* Operating System and version: MacOS 14.5
* Kotlin Version: 2.0.20-RC

**Comment by 3flex** [comment]
Are you using a snapshot versions of the Gradle plugin?

Are you using a custom detekt task, or one that was created by the Gradle plugin?

**Comment by 3flex** [comment]
Can you please provide either a link to a Gradle  build scan or provide a reproducer?

**Comment by 3flex** [comment]
Thanks - the reproducer appears to show these issues:
```
/home/runner/work/detekt-7531/detekt-7531/src/commonMain/kotlin/bug/Bug.kt:7:1: error: 'expect' and 'actual' declarations can be used only in multiplatform projects. Learn more about Kotlin Multiplatform: https://kotl.in/multiplatform-setup
expect fun myExpectFun()
^^^^^^
/home/runner/work/detekt-7531/detekt-7531/src/jvmMain/kotlin/bug/Bug.kt:7:1: error: 'expect' and 'actual' declarations can be used only in multiplatform projects. Learn more about Kotlin Multiplatform: https://kotl.in/multiplatform-setup
actual fun myExpectFun() = Unit
^^^^^^
```

That appears to be a valid issue but it's not a reproducer for the issue you originally mentioned (`error: unresolved reference`). Can you please provide a reproducer for your original issue? Enable debug mode in detekt to confirm that the messages are the same.

I've opened a new issue for the multiplatform feature not being enabled when detekt runs as this needs to be fixed: https://github.com/detekt/detekt/issues/7533

**Comment by 3flex** [comment]
We'll need a way to execute compiler plugins before/during detekt's analysis of the code.

I haven't used the serialization plugin but the problem seems to be caused by the compiler plugin generating code that is (currently) invisible to the detekt analysis tasks: https://github.com/Kotlin/kotlinx.serialization/blob/v1.7.1/docs/serializers.md#plugin-generated-serializer

> Every class marked with the @Serializable annotation, like the Color class from the previous example, gets an instance of the [KSerializer](https://kotlinlang.org/api/kotlinx.serialization/kotlinx-serialization-core/kotlinx.serialization/-k-serializer/index.html) interface automatically generated by the Kotlin Serialization compiler plugin. We can retrieve this instance using the .serializer() function on the class's companion object.

I'll update the title of this issue. If there other, unique issues you've found, can you please open new issues for those?

This feedback & testing is extremely valuable to find these gaps, so thank you for trying things out!

**Comment by bddckr** [comment]
Just chiming in here with another data point. In my Gradle build with 100+ projects, of the TODO compiler errors detekt reports, 100% are for `.serializer()` generated by Kotlin Serialization. There are a few errors from something else at first glance, but turns out those are due to the type checking failing earlier due to the serializer problem (which makes sense).

**Comment by bddckr** [comment]
I was reminded that Metro had to solve an ordering problem with the Compose compiler plugin, which we can see [here](https://github.com/ZacSweers/metro/blob/e2cf07583373e4fb7d0b7271a259da430deccdb1/gradle-plugin/src/main/kotlin/dev/zacsweers/metro/gradle/MetroGradleSubplugin.kt#L194-L201). That leads us to the official docs [here](https://kotlinlang.org/docs/compiler-reference.html#xcompiler-plugin-order-plugin-before-plugin-after).

That seems like what would solve this, if the detekt compiler plugin were to apply it similarly like Metro does, but instead for the Kotlin Serialization plugin ID.

That said, I tried to workaround this myself in my project for now via the following:
```kotlin
// https://github.com/detekt/detekt/issues/7531
extensions.configure<KotlinBaseExtension> {
    @Suppress("UNCHECKED_CAST")
    this as HasConfigurableKotlinCompilerOptions<KotlinJvmCompilerOptions>

    compilerOptions {
        freeCompilerArgs.add(
            "-Xcompiler-plugin-order=org.jetbrains.kotlinx.serialization>detekt-compiler-plugin"
        )
    }
}
```
I'm doing that in my convention plugin that applies and configures detekt's Gradle plugin.

I tried `org.jetbrains.kotlinx.serialization>detekt-compiler-plugin` and `org.jetbrains.kotlinx.serialization>io.github.detekt.detekt-compiler-plugin`, but neither of the two seems to help. I wonder whether that means this whole suggestion I'm raising here isn't a real solution, or whether I'm just doing something wrong. 🤔 My compiler plugin knowledge is pretty much non-existent, so please take this with a grain of salt.

**Comment by 3flex** [comment]
The original issue was about detekt Gradle tasks rather than the compiler plugin. There may be different solutions for each, not sure yet.

The (experimental!) compiler plugin has not been updated for a while so I'd recommend switching to use the Gradle plugin at this point.

**Comment by bddckr** [comment]
I'm applying the Gradle plugin only.

Isn't what I'm theorizing above applying to this issue nonetheless? The compiler plugin is applied by default, isn't it? 🤔 
https://github.com/detekt/detekt/blob/171abcbcc64b1773de1c48aedeca4b40f94b4840/detekt-gradle-plugin/src/main/kotlin/dev/detekt/gradle/plugin/DetektBasePlugin.kt#L143

**Comment by BraisGabin** [comment]
I was able to create a functional test to reproduce this issue. You can see it at: 0e2cf2685f5c10d9c7dbb236553bd4db1a651299

I'm not 100% sure why we are able to see types that aren't included #9082 but we aren't able to do that with the things created by a compiler plugin. The `serializer()` function is in the classpath.

**Comment by BraisGabin** [comment]
🤯🤯🤯 c3d6ca1a7c351f4a35b7a703820ba8ca6b8927a7. If I move the `@Serializable` class to an excluded file then everything works. I guess that Analyzer API uses the classpath in this case because the class `Book` isn't defined in the files it knows.

**Comment by FletchMcKee** [comment]
@3flex I may have this partly solved, but it's definitely a WIP. I think you're correct that we'll need to wire `pluginClasspath` and `pluginOptions` to the Detekt tasks, and then wire all of that to EnvironmentAware.

[KotlinCompilerPluginsProvider](https://github.com/JetBrains/kotlin/blob/1d6b471bbc6a1128df5c4d69cf39b451482a9ebc/analysis/analysis-api-platform-interface/src/org/jetbrains/kotlin/analysis/api/platform/projectStructure/KotlinCompilerPluginsProvider.kt#L16-L23) appears to be key for exposing the compiler plugins to the analysis API engine. `StandaloneAnalysisAPISessionBuilder` has a `registerCompilerPluginServices` method that handles this for us.

However the compiler plugins still need to be registered in the CompilerConfiguration for this to work. I think [PluginCliParser.loadPluginsSafe](https://github.com/JetBrains/kotlin/blob/ef3955975a5311f82a1c13acde9b98e1dbc1de41/compiler/cli/src/org/jetbrains/kotlin/cli/jvm/plugins/PluginCliParser.kt#L92-L122) might be the way to do this, but maybe there's a better way. I haven't figured out the right way to add `pluginOptions`, and that method also expects `pluginConfigurations` and `pluginOrderConstraints`, so we probably have to figure out the proper way of setting all of these. 

I added the changes @BraisGabin made in 0e2cf2685f5c10d9c7dbb236553bd4db1a651299 to my WIP to verify this. These are the `pluginClasspath` jars added in JvmSpec:
```
kotlin-scripting-compiler-embeddable-2.3.10.jar
kotlin-serialization-compiler-plugin-embeddable-2.3.10.jar
kotlin-scripting-compiler-impl-embeddable-2.3.10.jar
kotlin-scripting-jvm-2.3.10.jar
kotlin-scripting-common-2.3.10.jar
kotlin-stdlib-2.3.10.jar
kotlin-script-runtime-2.3.10.jar
annotations-13.0.jar
```
Initially I tried adding these without any `pluginOptions` and it fails with a ClassNotFoundException due to these script plugins. I was looking at how Kotlin does this and noticed we could disable the script logic by adding ["plugin:kotlin.scripting:disable=true"](https://github.com/JetBrains/kotlin/blob/60a136bdedf38876ceeb2c3b5e951f46d89aff41/compiler/cli/src/org/jetbrains/kotlin/cli/common/CLICompiler.kt#L250) as a pluginOption. Doing this did allow the test to pass, so I think it's a start. I can create a branch on my fork with these changes if you'd like to get a better idea of what changed. I'll highlight the main changes outside of all of the wiring:

EnvironmentAware:
```kotlin
init {
    buildStandaloneAnalysisAPISession(disposable) {
        registerCompilerPluginServices(configuration)
        ...
    }
}
```
KotlinEnvironmentUtils:
```kotlin
fun createCompilerConfiguration(
    ...
    compilerPluginClasspath: String?,
    disposable: Disposable,
): CompilerConfiguration {
    ...
    return CompilerConfiguration().apply {
        ...
        compilerPluginClasspath?.split(File.pathSeparator)
            .orEmpty()
            .takeIf { it.isNotEmpty() }
            ?.let { pluginClasspaths ->
                PluginCliParser.loadPluginsSafe(
                    pluginClasspaths = pluginClasspaths,
                    pluginOptions = listOf("plugin:kotlin.scripting:disable=true"),
                    pluginConfigurations = emptyList(),
                    pluginOrderConstraints = emptyList(),
                    configuration = this,
                    parentDisposable = disposable,
                )
            }
    }
}
```

**Comment by FletchMcKee** [comment]
Here's a [comparison](https://github.com/detekt/detekt/compare/main...FletchMcKee:detekt:compiler-plugin-support) of that branch and main.

Edit: The above works for the kotlinx serialization compiler plugin, but doesn't work with the compose compiler plugin. The compose compiler plugin [relocates all `com.intellij` classes](https://github.com/JetBrains/kotlin/blob/master/plugins/compose/compiler/build.gradle.kts#L25-L27) and uses an embedded kotlin-compiler. Not sure how feasible my approach above is, I imagine a number of compiler plugins could do something similar.

**Comment by damianw** [comment]
To add some preliminary support, could this get started by just adding support in the CLI (and leave Gradle to be figured out later)? That would be enough for our use case, I think.

From @FletchMcKee's POC, the main changes needed appear to be:
- Additional options in the CLI
- Additional options in `CompilerSpec`
- Support in `EnvironmentAware`

Looking at `loadPluginsSafe` I guess there are two ways of specifying compiler plugins at this point - either classpath+options as separate items ("legacy" style), or combined `classpath1.jar,classpath2.jar=option1=value1,option2=value2` ("modern" style). I don't think both would be needed, so this change would require only one option to be added to the CLI, e.g. repeatable `--compiler-plugin-configuration`.

I'd be happy to put up a PR for this if it sounds reasonable.

**Comment by 3flex** [comment]
I'm declaring this issue blocked on #9676 and closing my PR for now.

See commit message in https://github.com/JetBrains/kotlin/commit/8cac77f16a987126772ff83695b579a7a6aae3c7. Based on this:
1. We can only expect official compiler plugins to work reliably with the Analysis API as they're sensitive to the compiler version used by AA.
2. The compiler version used by official Kotlin Analysis API packages is not necessarily the same as a production compiler version. This means third party compiler plugins that are not forward or backward compatible with different compiler versions are not likely to work. Even if they are, they may only be tested against official compiler releases, which we'll most likely be decoupled from in future. Some info from a Kotlin maintainer [here](https://youtrack.jetbrains.com/issue/KT-88573/Analysis-API-Standalone-kotlin-compiler-for-ide-artifacts-are-classpath-incompatible-since-2.4.20-Beta1#focus=Comments-27-14292610.0-0) and [here](https://youtrack.jetbrains.com/issue/KT-88573/Analysis-API-Standalone-kotlin-compiler-for-ide-artifacts-are-classpath-incompatible-since-2.4.20-Beta1#focus=Comments-27-14313121.0-0).
3. The IDE substitutes plugins on the compile classpath with the compiler versions compatible with the current Analysis API version when it does its analysis in the IDE. We should do the same thing to align behaviour.

From here, we need to switch to the official AA packages where available, add support for official compiler plugins using the same substitution mechanism as the IDE so we're not deviating from that behaviour, and then include experimental, best-effort support for third party plugins (not sure which third party plugins are most popular, but perhaps we test Metro, Arrow and Poko?).
